---
title: "MOFA+: integration of a heterogeneous time-course single-cell RNA-seq dataset."
author:
  name: "Ricard Argelaguet"
  affiliation: "European Bioinformatics Institute, Cambridge, UK"
  email: "ricard@ebi.ac.uk"
date: "`r Sys.Date()`"
output:
  BiocStyle::html_document:
    toc: true
vignette: >
  %\VignetteIndexEntry{scRNA}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

# Description

This vignette demonstrates the multi-group integration framework of MOFA+ on a single data modality.

We consider a data set of scRNA-seq experiments where 16,152 cells were isolated from a total of 8 mouse embryos from developmental stages E6.5, E6.75, E7.0 and E7.25 (two embryos per stage), spanning post-implantation and early gastrulation.  
Despite differences in developmental time, embryos are expected to contain similar subpopulations of cells. Hence, MOFA+ should detect the existence of biological sources of variation that are shared between groups. 

The data set we use here is a simplified subset version of the original data set, which can be visualised and downloaded from [here](https://marionilab.cruk.cam.ac.uk/MouseGastrulation2018/).

```{r global_options, include=FALSE}
knitr::opts_chunk$set(warning=FALSE, message=FALSE, fig.align="center")
```

# Load libraries

Load dependencies. Make sure that `MOFA2` is imported last, to avoid collisions with functions from other packages
```{r, echo=FALSE}
library(Seurat)
library(ggplot2)
library(MOFA2)
```

Define cell type colors for the visualisations
```{r}
colors <- c(
  "Epiblast" = "grey70",
  "Primitive Streak" = "sandybrown",
  "Mesoderm" = "violetred",
  "ExE endoderm" = "#548B54",
  "ExE ectoderm" = "black"
)
```

# Load data

The RNA expression data has been processed using Seurat. It has already been normalised and subset to the top 5,000 most variable genes (after regressing out the group effect).

```{r}
# load("/Users/ricard/data/mofa2_vignettes/gastrulation10x_seurat.RData")
load(url("ftp://ftp.ebi.ac.uk/pub/databases/scnmt_gastrulation/mofa2/scrna_gastrulation/gastrulation10x_seurat.RData"))
```

# Create MOFA object

Group cells according to the embryo and the stage they comne from
```{r}
seurat@meta.data$stage_sample <- paste(seurat@meta.data$stage,seurat@meta.data$sample, sep="_")
unique(seurat@meta.data$stage_sample)
```

```{r}
MOFAobject <- create_mofa(seurat, groups = "stage_sample")
```

## Plot data overview
```{r}
plot_data_overview(MOFAobject)
```

## Define MOFA options
Data options: let's use default
```{r}
data_opts <- get_default_data_options(MOFAobject)
```

Model options: let's use default
```{r}
model_opts <- get_default_model_options(MOFAobject)
```

Training options
```{r}
train_opts <- get_default_training_options(MOFAobject)
train_opts$convergence_mode <- "fast"
train_opts$seed <- 42
```

## Prepare the MOFA object
```{r}
MOFAobject <- prepare_mofa(
  object = MOFAobject,
  data_options = data_opts,
  model_options = model_opts,
  training_options = train_opts
)
```


# Train the MOFA model
This step can take quite some time (~2h with standard CPU inference), so we provide a pre-trained model in the next chunk of code.  

Note: if you train the model from scratch, the results will not be 100% reproducible with the pre-trained model (but hopefully similar enough!).

```{r, eval=FALSE}
outfile = paste0(getwd(),"/model.hdf5")
MOFAmodel <- run_mofa(MOFAobject, outfile = outfile)
```

## Load pre-computed model

MOFA models are saved in hdf5 format and can be loaded into R with the function `load_model`. In this case, however, we provide the trained model as an `RData` file
```{r}
# MOFAmodel <- load_model(outfile)

# load("/Users/ricard/data/mofa2_vignettes/gastrulation10x_mofa.RData")
load(url("ftp://ftp.ebi.ac.uk/pub/databases/scnmt_gastrulation/mofa2/scrna_gastrulation/gastrulation10x_mofa.RData"))
```

# Inspect MOFA object
The MOFAobject consists of multiple slots where relevant data and information is stored. For descriptions, you can read the documentation by `?MOFA`
```{r}
slotNames(MOFAmodel)
```

## Cell metadata:  

* *stage*: developmental stage
* *lineage*: cell type annotation
* *group*: the grouping used for MOFA, corresponding to stage + embryo 
```{r}
head(MOFAmodel@samples_metadata)
```

## Overview of training data
The function `plot_data_overview` can be used to obtain an overview of the input data. 
It shows how many views (rows) and how many groups (columns) exist, what are their corresponding dimensionalities and how many missing information they have (grey bars).  
In this case we have one view (RNA expression, a total of 5,000 genes) and 6 groups that correspond to different embryos at different stages of development, for a total of 16,152 cells.
```{r}
plot_data_overview(MOFAmodel, colors = c("RNA"="darkgreen"))
```

# Plot variance explained

Quantifying the variance explained per factor across groups and views is probably the most important plot that MOFA+ generates. It summarises the (latent) signal from a complex heterogeneous data set in a single figure. 

## Plot variance explained per factor across groups
```{r}
plot_variance_explained(MOFAmodel, x="group", y="factor")
```

There is a lot of information contained in this plot. Factor 1 and Factor 2 explain a lot of variance across multiple groups. In contrast, Factor 4 increases its activity from E6.5 to E7.5, indicating that it captures a source of variation that emerges at E6.5.  

To make the plot less dominated by the top 2 factors, one can scale the R2 values with the arguments `min_r2` and `max_r2`:
```{r}
plot_variance_explained(MOFAmodel, x="group", y="factor", max_r2 = 0.15)
```

We can also plot the total variance explained per group (with all factors) by adding the argument `plot_total = TRUE`. Notably, only 10 factors are sufficient to capture between 35% and 55% of the transcriptional variance per embryo

## Plot total variance explained per group

```{r}
p <- plot_variance_explained(MOFAmodel, x="group", y="factor", plot_total = T)

p[[2]]
```

## Plot variance explained for individual features

We can also inspect the variance explained by the MOFA factors for individual features. This is useful to disentangle how much of the variation of a gene is "noise" (high variance but low $R^2$) versus coordinated variance (high $R^2$)

```{r}
features <- c("Rbp4","Ttr","Spink1","Mesp1","E130311K13Rik","Hey1")
```

Variance explained by all factors across all groups
```{r}
plot_variance_explained_per_feature(
  MOFAmodel, 
  factors = "all",
  groups = "all",
  view = "RNA",
  features = features
)
```

# Characterise Factor 1

## Visualisation of factor values

Each factor ordinates cells along a one-dimensional axis that is centered at zero. Samples with different signs indicate opposite phenotypes, with higher absolute value indicating a stronger phenotype. For example, if the $k$-th factor captures the variability associated with cell cycle, we could expect cells in Mitosis to be at one end of the factor (irrespective of the sign, only the relative positioning being of importance). In contrast, cells in G1 phase are expected to be at the other end of the factor. Cells with intermediate phenotype, or with no clear phenotype (i.e. no cell cycle genes profiled), are expected to be located around zero.

Let's plot Factor 1 values and we color cells by lineage assignment. Clearly, this factors captures the emergence of ExE endoderm.

```{r}
plot_factor(MOFAmodel, 
  factor = 1,
  color_by = "lineage"  # lineage is a column in MOFAmodel@samples.metadata
) + scale_color_manual(values=colors)
```

Here are other ways of representing the same plot:
```{r}
p <- plot_factor(MOFAmodel, 
  factor = 1,
  color_by = "lineage",
  dot_size = 0.2,      # change dot size
  dodge = T,           # dodge points with different colors
  legend = F,          # remove legend
  add_violin = T,      # add violin plots,
  violin_alpha = 0.25  # transparency of violin plots
)

p <- p + 
  scale_color_manual(values=colors) +
  scale_fill_manual(values=colors)

p
```

One can also change the default groups by some manually defined grouping structure. For example a column in the sample metadata
```{r}
plot_factor(MOFAmodel, 
  factor = 1,
  color_by = "lineage",
  group_by = "stage",         # cells are now grouped by stage, rather than stage+embryo,
  legend = F
) + scale_color_manual(values=colors)
```

Combinations of factors can be plotted with `plot_factors`:
```{r}
plot_factors(MOFAmodel, 
  factors = c(1,4),
  color_by = "lineage"
) + scale_color_manual(values=colors)
```

## Visualisation of weights

The weights provide a score for each gene on each factor. Genes with no association with the factor are expected to have values close to zero, whereas genes with strong association with the factor are expected to have large absolute values. The sign of the loading indicates the direction of the effect: a positive loading indicates that the feature is more active in the cells with positive factor values, and viceversa. \\
Following the cell cycle example from above, we expect genes that are upregulated in the M phase to have large positive weights, whereas genes that are downregulated in the M phase (or, equivalently, upregulated in the G1 phase) are expected to have large negative weights.\\

Let's plot the distribution of weights for Factor 1.
```{r, warnings=FALSE, message=FALSE}
plot_weights(MOFAmodel,
  view = "RNA",
  factor = 1,
  nfeatures = 10,     # Top number of features to highlight
  scale = T           # Scale weights from -1 to 1
)
```

If we are not interested in the directionality of the effect, we can take the absolute value of the weights (`abs=TRUE`). We can also highlight some genes of interest using the argument `manual` to see where in the distribution they lie:
```{r}
plot_weights(MOFAmodel,
  view = "RNA",
  factor = 1,
  nfeatures = 5,
  manual = list(c("Snai1","Mesp1","Phlda2"), c("Rhox5","Elf5")),
  color_manual = c("darkgreen","red"),
  scale = T,
  abs = T
)
```

If you are not interested in the full distribution, but just on the top weights, you can do:
```{r}
plot_top_weights(MOFAmodel, 
  view = "RNA", 
  factor = 1, 
  nfeatures = 10,
  scale = T, 
  abs = T
)
```

## Visualisation of factors coloured by gene expression

We expect negative weights For Factor 1 to be marker genes of ExE Endoderm. If we plot Factor 1 colouring cells by gene expresion of the top genes:
```{r}
genes <- c("Ttr","Apom","Apoa1","Amn")

for (i in genes) {
  
  p <- plot_factor(MOFAmodel,
    factor = 1,
    color_by = i
  ) + scale_colour_gradientn(colours = terrain.colors(10)) # change color scale
  
  print(p)
  
}
```

The weights are useful to get an idea of which are top genes that drive the factors. However, to get an idea of how well Factors are associated with genomic features we can generate a scatter plot of the Factor values against gene expression for the genes with the highest weights:

Positive weights
```{r}
p <- plot_data_scatter(MOFAmodel, 
  view = "RNA", 
  factor = 1, 
  features = 6,         # Number of features to show
  sign = "positive",     # select top 6 features with positive weights
  color_by = "lineage",  # color cells by lineage
  add_lm = TRUE          # add linear regression estimates
)
p <- p + scale_color_manual(values=colors) + theme(legend.position = "none")

print(p)
```

Negative weights
```{r}
p <- plot_data_scatter(MOFAmodel, 
  view = "RNA", 
  factor = 1, 
  features = 6, 
  sign = "negative",
  color_by = "lineage",
  add_lm = TRUE
)
p <- p + scale_color_manual(values=colors) + theme(legend.position = "none")

print(p)
```


# Non-linear dimensionality reduction from the MOFA factors

The latent space inferred by MOFA can be employed as input to other single-cell algorithms that learn non-linear manifolds such as UMAP or t-SNE. This can be very useful to identify cellular populations and reconstruct complex pseudotime trajectories.  

Run t-SNE (takes ~2min):
```{r}
set.seed(42)
MOFAmodel <- run_tsne(MOFAmodel)
```

In this data set, we see that the combination of MOFA factors have enough information to discriminate all cell types. 

```{r }
plot_dimred(MOFAmodel,
  method = "TSNE",
  color_by = "lineage"
) + scale_color_manual(values=colors)
```

This is not surprising and yields a similar result as running PCA (after batch correction). Let's move to more exciting applications in the next vignettes!

# sessionInfo

```{r}
sessionInfo()
```

