<!DOCTYPE html>
<html lang="en-US">
  <head>

    <meta charset="UTF-8">

<!-- Begin Jekyll SEO tag v2.6.1 -->
<title>Tutorials | Multi-Omics Factor Analysis (MOFA+)</title>
<meta name="generator" content="Jekyll v3.9.0" />
<meta property="og:title" content="Tutorials" />
<meta property="og:locale" content="en_US" />
<meta name="description" content="Multi-Omics Factor Analysis V2 (MOFA+)" />
<meta property="og:description" content="Multi-Omics Factor Analysis V2 (MOFA+)" />
<link rel="canonical" href="http://localhost:4000/tutorials.html" />
<meta property="og:url" content="http://localhost:4000/tutorials.html" />
<meta property="og:site_name" content="Multi-Omics Factor Analysis (MOFA+)" />
<script type="application/ld+json">
{"headline":"Tutorials","description":"Multi-Omics Factor Analysis V2 (MOFA+)","@type":"WebPage","url":"http://localhost:4000/tutorials.html","@context":"https://schema.org"}</script>
<!-- End Jekyll SEO tag -->

    <meta name="viewport" content="width=device-width, initial-scale=1">
    <meta name="theme-color" content="#157878">
    <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
    <link rel="stylesheet" href="/assets/css/style.css?v=33b8e818418c8e3e88668fb03d1d52bc031c390a">
  </head>
  <body>
    <!-- <a id="skip-to-content" href="#content">Skip to the content.</a> -->

    <header class="page-header" role="banner">
      <h1 class="project-name">Tutorials</h1>
      <h2 class="project-tagline">Multi-Omics Factor Analysis V2 (MOFA+)</h2>

      <a href="index.html" class="btn">Home</a>
      <a href="installation.html" class="btn">Installation</a>
      <a href="tutorials.html" class="btn">Tutorials</a>
      <a href="http://www.ebi.ac.uk/shiny/mofa/" class="btn">Interactive web server</a>
      <a href="faq.html" class="btn">FAQ</a>
      <a href="troubleshooting.html" class="btn">Troubleshooting</a>
      <a href="contact.html" class="btn">Contact</a>
      
        <a href="https://github.com/bioFAM/MOFA2" class="btn">View on GitHub</a>
      
    </header>

    <main id="content" class="main-content" role="main">
      <h2 id="learning-the-basics">Learning the basics</h2>

<ul>
  <li><a href="https://github.com/bioFAM/MOFA2/blob/master/MOFA2/vignettes/getting_started.md"><strong>Getting started</strong></a>: general overview and description of the method, for both R and Python users.</li>
</ul>

<h3 id="r">R</h3>
<ul>
  <li>
    <p><a href="https://raw.githack.com/bioFAM/MOFA2/master/MOFA2/vignettes/getting_started_R.html"><strong>Training a model in R</strong></a>: using simple simulated data</p>
  </li>
  <li>
    <p><a href="https://raw.githack.com/bioFAM/MOFA2/master/MOFA2/vignettes/downstream_analysis.html"><strong>Downstream analysis in R</strong></a>: using simple simulated data</p>
  </li>
  <li>
    <p><a href="https://raw.githack.com/bioFAM/MOFA2/master/MOFA2/vignettes/GSEA.html"><strong>Gene set enrichment analysis</strong></a>: demonstrates how to do gene set enrichment analysis in R.</p>
  </li>
  <li>
    <p><a href="https://raw.githack.com/bioFAM/MOFA2/master/MOFA2/vignettes/stochastic_inference.html"><strong>Demonstration of the stochastic inference algorithm</strong></a>: this is only useful for very large data sets and when having access to GPUs.</p>
  </li>
</ul>

<h3 id="python">Python</h3>

<ul>
  <li>
    <p><a href="https://github.com/bioFAM/MOFA2/blob/master/mofapy2/notebooks/getting_started_python.ipynb"><strong>Training a model in Python</strong></a>: a jupyter notebook demonstrating how to train a MOFA model using simple simulated data</p>
  </li>
  <li>
    <p><strong>Downstream analysis in Python</strong>: in preparation…</p>
  </li>
</ul>

<h2 id="case-examples-using-real-data-in-r">Case examples using real data (in R)</h2>

<ul>
  <li>
    <p><a href="https://raw.githack.com/bioFAM/MOFA2/master/MOFA2/vignettes/CLL.html"><strong>(authors’ favourite) Analysis of chronic lymphocytic leukaemia cohort for personalised medicine</strong></a>: a bulk multi-omics data set. <a href="https://www.embopress.org/doi/full/10.15252/msb.20178124#msb178124-fig-0002">Figure 2 and 3 of the MOFA v1 paper</a>.</p>
  </li>
  <li>
    <p><a href="https://raw.githack.com/bioFAM/MOFA2/master/MOFA2/vignettes/10x_scRNA_scATAC.html"><strong>(authors’ favourite) Integrative analysis of the Chromium Single Cell Multiome ATAC + Gene Expression assay</strong></a>: this is the result of a collaboration between the MOFA team and the 10x Genomics R&amp;D team to provide a downstream analysis pipeline for the new RNA+ATAC multi-modal technology.</p>
  </li>
  <li>
    <p><a href="https://raw.githack.com/bioFAM/MOFA2/master/MOFA2/vignettes/scRNA_gastrulation.html"><strong>Analysis of a time course scRNA-seq data set using the multi-group framework</strong></a>: <a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-020-02015-1#Fig2">Figure 2 of the MOFA+ paper</a>. Demonstrates the multi-group functionality and how to train a MOFA model from a Seurat object.</p>
  </li>
  <li>
    <p><a href="https://raw.githack.com/bioFAM/MOFA2/master/MOFA2/vignettes/scNMT_gastrulation.html"><strong>Integration of scNMT-seq data  (single-cell multi-omics)</strong></a>: <a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-020-02015-1#Fig4">Figure 4 of the MOFA+ paper</a>. Demonstrates the simultaneous multi-view and multi-group functionality using the <a href="https://www.nature.com/articles/s41586-019-1825-8">scNMT-seq mouse gastrulation atlas</a>.</p>
  </li>
  <li>
    <p><a href="https://raw.githack.com/bioFAM/MOFA2/master/MOFA2/vignettes/SNARE_seq.html"><strong>Integration of SNARE-seq data (single-cell multi-omics)</strong></a>. Demonstrates how MOFA can be used for the analysis of paired scRNA+scATAC data (from the same cell) using a multi-modal Seurat object. This data set is very noisy and the results are not fantastic, we suggest you have a look at the Chromium Single Cell Multiome ATAC + Gene Expression vignette instead.</p>
  </li>
  <li>
    <p><a href="https://raw.githack.com/bioFAM/MOFA2/master/MOFA2/vignettes/microbiome_vignette.html"><strong>Analysis of multi-modal microbiome data</strong></a>: we demonstrate how to systematically integrate viral, fungal and bacterial sequence data.
*
<!-- * [**Robustness analysis and model selection**](https://raw.githack.com/bioFAM/MOFA2/master/MOFA2/vignettes/model_selection.html) --></p>
  </li>
</ul>

<!-- * [**Analysis of single-cell DNA methylation data (in R)**](https://github.com/bioFAM/MOFA2/blob/master/MOFA2/vignettes/scMethylation_cortex.html): Figure 3 of the paper, in preparation... -->

<ul>
  <li><strong>Analysis of CITE-seq data</strong>: still in preparation, reach us if you have questions…</li>
</ul>

<h2 id="case-examples-in-python">Case examples (in Python)</h2>

<p>In preparation…</p>

<h2 id="template-scripts">Template scripts</h2>

<p>We provide two template scripts to train your model in <a href="https://github.com/bioFAM/MOFA2/blob/master/template_script.R">R</a> and <a href="https://github.com/bioFAM/MOFA2/blob/master/template_script.py">Python</a>. Please see the tutorials for more details regarding the input data format, etc.</p>

<h2 id="documentation">Documentation</h2>

<ul>
  <li><a href="https://github.com/bioFAM/MOFA2/blob/master/documentation.pdf">R package</a></li>
  <li><a href="">Python package</a></li>
</ul>


<!--       <footer class="site-footer">
        
          <span class="site-footer-owner"><a href="https://github.com/bioFAM/MOFA2">MOFA2</a> is maintained by <a href="https://github.com/bioFAM">bioFAM</a>.</span>
        
        <span class="site-footer-credits">This page was generated by <a href="https://pages.github.com">GitHub Pages</a>.</span>
      </footer> -->

    </main>
  </body>
</html>
